### PR TITLE
strengthen doc warning about CString::from_raw

### PR DESCRIPTION
**Pull request by durka**
Saw unsound code using this function on IRC.

### CODE REVIEW

File: src/libstd/ffi/c_str.rs [hunk 1]
```diff
@@ -225,8 +225,11 @@ impl CString {
     /// Retakes ownership of a `CString` that was transferred to C.
     ///
     /// This should only ever be called with a pointer that was earlier
-    /// obtained by calling `into_raw` on a `CString`. Additionally, the length
-    /// of the string will be recalculated from the pointer.
+    /// obtained by calling `into_raw` on a `CString`. In particular, using this method
+    /// to create a `CString` pointing at memory that will be freed by other code
+    /// (such as a C library) will lead to undefined behavior!
```
Comment (nagisa) [inline src/libstd/ffi/c_str.rs]: Creating `CString` from arbitrary allocation is not UB. Doing anything other than `mem::forget(create)` and `CString::from_raw(create)` what is.

Comment (nagisa) [inline src/libstd/ffi/c_str.rs]: I feel like we’ve brainstormed a nice way to say this somewhere already. The way `String` and `Vec` state this invariant is pretty good.

**Comment by steveklabnik** [comment]
@bors: r-

sure, let's talk about it :smile:

**Comment by jonathandturner** [comment]
PS: also this fails tidy.

File: src/libstd/ffi/c_str.rs (hunk 1 above)
Comment (durka) [inline src/libstd/ffi/c_str.rs]: You're right. That's why I didn't say "is UB". I said "will lead to UB" which is indeed likely if you use `from_raw` on anything that didn't come from `into_raw`.

**Comment by jonathandturner** [comment]
> screw you, tidy

Hmmm... want to squash those commits? ;)

File: src/libstd/ffi/c_str.rs (hunk 1 above)
Comment (nagisa) [inline src/libstd/ffi/c_str.rs]: Well, sure, it doesn’t always lead to UB regardless of its arguments (i.e. if you immediately forget it, however unreasonable such code would be). Moreover using similar format as `String::from_raw` and `Vec::from_raw` helps consistency :)

**Comment by durka** [comment]
@jonathandturner sure I will squash if this goes in as is. Sounds like it needs some more work. I can copy the style from [Vec::from_raw_parts](https://doc.rust-lang.org/beta/std/vec/struct.Vec.html#method.from_raw_parts) as @nagisa suggested.

**Comment by steveklabnik** [comment]
Yes, let's copy that style and get this PR merged, finally.

**Comment by durka** [comment]
Done.

**Pull request merged**